src/main/java/com/puppycrawl/tools/checkstyle/FileStatefulModule.java

```diff
@@ -0,0 +1,34 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2017 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle;
+
+/**
+ * This interface means that the module contains file-related context and therefore
+ * cannot be used from the separate threads at the same time.
+ * It is guaranteed that the whole file will be processed inside the same thread.
+ * It is not guaranteed that each file will have it's own thread - there might be a list of files,
+ * which will be executed on the same thread.
```

Comment:
Should we mention the instance somewhere in this along with the thread?
Each file gets a brand new instance, right?

Reply:
> Each file gets a brand new instance, right?

@rnveach  No, actually, there is only N checks clones, where N is the number of threads.

Reply:
Added notes about instances